api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java

```diff
@@ -1096,4 +1105,84 @@ public DbUser confirmPublications() {
         dbUser,
         Agent.asUser(dbUser));
   }
+
+  /** Send an Access Renewal Expiration or Warning email to the user, if appropriate */
+  @Override
+  public void maybeSendAccessExpirationEmail(DbUser user) {
+    // TODO combine with CT expiration logic when available to send AT MOST ONE email
+
+    final Optional<Timestamp> rtExpiration = getRegisteredTierExpirationForEmails(user);
+    rtExpiration.ifPresent(expiration -> maybeSendRegisteredTierExpirationEmail(user, expiration));
+  }
+
+  /**
+   * Return the user's registered tier access expiration time, for the purpose of sending an access
+   * renewal reminder or expiration email.
+   *
+   * <p>First: ignore any bypassed modules. These are in compliance and do not need to be renewed.
+   *
+   * <p>Next: do all un-bypassed modules have expiration times? If yes, return the min (earliest).
+   * If no, either the feature flag is not set or the user does not have access for reasons other
+   * than access renewal compliance. In either negative case, we should not send an email.
+   *
+   * <p>Note that this method may return EMPTY for both valid and invalid users, so this method
+   * SHOULD NOT BE USED FOR ACCESS DECISIONS.
+   */
+  private Optional<Timestamp> getRegisteredTierExpirationForEmails(DbUser user) {
```

Comment:
the logic here is a little fiddly.  Please review as closely as possible, and LMK if there are any additional comments or tests that would clarify this.